saga/Repository.py

```diff
@@ -381,4 +385,113 @@ def _copy_file_to_dir(self, src, dst):
         if not os.path.isdir(dirname):
             os.makedirs(dirname)
 
-        shutil.copyfile(src, join(dst, src))
\ No newline at end of file
+        shutil.copyfile(src, join(dst, src))
+
+
+
+    def pull_folder(self, relative_folder_path):
+        # api-endpoint 
+        URL = "http://localhost:3000/get-folder"
+
+        base_directory = self.base_directory
+
+        shutil.rmtree(self.saga_directory)
+                
+        if not base_directory[len(base_directory) -1] == '/':
+            base_directory = base_directory + "/"
+        
+        try:
+            response = requests.get(url = URL, data={'folder_location' : relative_folder_path})
+            paths = response.json()
+            
+            # download all of the files
+            for path in paths['file_paths']:
+                self._pull_file(path)
+
+            # make all of the empty folders
+            for path in paths['empty_folder_paths']:
+                empty_folder_location = base_directory + path
+                if os.path.exists(empty_folder_location):
+                    os.rmdir(empty_folder_location)
+                os.makedirs(empty_folder_location)
+                print("Downloaded Empty Folder: {}".format(empty_folder_location))
+
+
+        except:
+            print("response failed")
+
+    def _pull_file(self, relative_file_path):
+
+        # api-endpoint 
+        URL = "http://localhost:3000/download"
+        
+        # location given here 
+        file_location = self.base_directory
+
+        if not file_location[len(file_location) -1] == '/':
+            file_location = file_location + "/"
+                
+        # append the file name onto the path
+        file_location = file_location + relative_file_path
+
+        file_name = relative_file_path.split('/')
```

Comment:
There are os functions that do all this work for you. Specifically, I think you are looking for: `os.path.basename` and `os.path.dirname`. Look em up!